### PR TITLE
deprecate std/mersenne

### PR DESCRIPTION
**Pull request by qaziquza**
I added a couple of functions to the Mersenne Twister module. I am sure that very few people use the module, so I picked it to be safe, as I am an incompetent algorithm writer.

### CODE REVIEW

**Comment by timotheecour** [comment]
* not DRY
* not generic
* not idiomatic
* comments like `## Might be useful at some point to someone.` (wrongly placed for docgen and, well, ...)

## what we need instead:
allowing reusing std/random with a generic interface allowing to plugin `MersenneTwister` or `sysrand` instead of the default random number generator

**Comment by qaziquza** [comment]
I meant "Replace lots of useless code with  a generic proc."

**Comment by timotheecour** [comment]
> I meant "Replace lots of useless code with a generic proc."

you can `git rebase -i` and force-push to change past commits

the problem with `proc sample*[T](mt: var MersenneTwister, arr: seq[T]): T =` is that it tries to mirror `random.sample`, and then soon enough we'll have to mirror the other procs too (`gauss` etc).

Clearly, this API mirroring is not a good design.

A good design would allow calling gauss, sample, etc generically on the random generator type

**Comment by konsumlamm** [comment]
> A good design would allow calling gauss, sample, etc generically on the random generator type

I feel like a RNG concept would be nice here (ofc that'd have to wait for concepts to stabilize), similar to Rust's [`Rng` trait](https://docs.rs/rand/0.8.4/rand/trait.Rng.html).

**Comment by Araq** [comment]
What's wrong with the default random number generator? I don't want interfaces everywhere, the parametrization either causes code bloat or indirect function calls...

**Comment by konsumlamm** [comment]
> What's wrong with the default random number generator? I don't want interfaces everywhere, the parametrization either causes code bloat or indirect function calls...

For one, the default RNG isn't cryptographically secure. Though I don't know why the `mersenne` module exists in the first place. But even for only two different RNGs, it would be nice to have a common interface, instead of manually duplicating all the functions, imo.

**Comment by Araq** [comment]
> But even for only two different RNGs, it would be nice to have a common interface, instead of manually duplicating all the functions, imo.

Well we can consider that once we have a second RNG that anybody out there cares about. :-)

**Comment by konsumlamm** [comment]
> > But even for only two different RNGs, it would be nice to have a common interface, instead of manually duplicating all the functions, imo.
> 
> Well we can consider that once we have a second RNG that anybody out there cares about. :-)

So if apparently noone cares about `std/mersenne` and `std/sysrand`, then why were they added in the first place?

**Comment by Araq** [comment]
> So if apparently noone cares about std/mersenne and std/sysrand, then why were they added in the first place?

mersenne is archaic and I don't know to which extend sysrand can implement a sane interface as there is some fallback to /dev/urandom (which is terrible).

**Comment by konsumlamm** [comment]
> mersenne is archaic

Hmm, perhaps we should just deprecate it then (and recommend `random` instead)?

**Comment by Vindaar** [comment]
> So if apparently noone cares about std/mersenne and std/sysrand, then why were they added in the first place?

I'm a physicist and care about it for example. If only, because in ROOT (CERN's HEP software framework) the class `TRandom3` is based on the Mersenne twister and is still very commonly used for Monte Carlo simulations in high energy physics.

**Comment by qaziquza** [comment]
Make it a nimble package? An alternative to the stdlib random? Those who want it can use it, and those who don't are not bothered?

**Comment by Vindaar** [comment]
> Make it a nimble package? An alternative to the stdlib random? Those who want it can use it, and those who don't are not bothered?

Not sure if you're replying to me or not.

But personally I don't care whether it lives in the stdlib or elsewhere. I care about the fact that the thing is implemented.
For sampling I use [alea](https://github.com/andreaferretti/alea/) already anyway, which precisely allows to wrap any RNG.

**Comment by Araq** [comment]
> Make it a nimble package? An alternative to the stdlib random? Those who want it can use it, and those who don't are not bothered?

Please make it a Nimble package.

**Comment by timotheecour** [comment]
> What's wrong with the default random number generator? I don't want interfaces everywhere, the parametrization either causes code bloat or indirect function calls...

there's little point in RT parametrization on the RNG type so let's just discuss the CT parametrization via eg

```nim
proc gauss(r: var Rand; mu = 0.0; sigma = 1.0): float
etc...

=>
type RngConcept* = concept
  proc next(a: var Self): uint64
  proc skipRandomNumbers(a: var Self) # could be optional

proc gauss(r: var RngConcept; mu = 0.0; sigma = 1.0): float
etc...
```

really not much difference complexity wise.

> I don't know to which extend sysrand can implement a sane interface as there is some fallback to /dev/urandom (which is terrible).

it wouldn't cause code bloat in the generated binary any more than what the program calls, so if you only instantiate it with only std/random's default RNG, it wouldn't increase binary size.

I'm not sure why it matters that it can fallback to /dev/urandom on some systems; such systems would anyway have that limitation if they wanted to implement `gauss` from scratch from urandom source of randomness. Note that urandom data can be grabbed in batches to avoid 1 call per requested value analog to how fread works via buffering

The point is we can have gauss, sample, etc available generically to std/sysrandom, the default RNG in std/random, as well as 3rd party libs that conform to `RngConcept`, for free.

**Comment by Araq** [comment]
> I'm not sure why it matters that it can fallback to /dev/urandom on some systems

It means that the operation can fail.

**Comment by qaziquza** [comment]
Should I remove or deprecate Mersenne, and create a new Github repo with Mersenne in it?

**Comment by timotheecour** [comment]
> It means that the operation can fail.

/dev/urandom (unlike /dev/random) is non-blocking, and what should be used as fallback (and that's indeed what we do). https://www.2uo.de/myths-about-urandom Is the best resource I found on this topic, and a good read.

If it fails (say with EINTR) we just retry, just like we do in io.readLine:
```nim
# in readLine:
    while true:
      # fixes #9634; this pattern may need to be abstracted as a template if reused;
      # likely other io procs need this for correctness.
      fgetsSuccess = c_fgets(addr line[pos], sp.cint, f) != nil
      if fgetsSuccess: break
      if errno == EINTR:
        errno = 0
        c_clearerr(f)
        continue
      checkErr(f)
      break
```

Again, the main point is that someone who wants to implement gauss/sample/uniform from scratch on top of a CSPRNG like system randomness would have no benefit in doing so over simply reusing existing code from std/random generalized as mentioned earlier to be generic on `RngConcept`.

**Comment by Araq** [comment]
> https://www.2uo.de/myths-about-urandom Is the best resource I found on this topic, and a good read.

It's irrelevant. You simply cannot read from `/dev/urandom` and know what you're gonna get, it might just be a file of this name. Yes, I know if this happens that the system is seriously compromised and there are lots of whack-a-mole solutions just so that we can continue to live in the dream land where pseudo-paths can be used instead of APIs. But the correct solution is to finally admit that Unix is a terrible design and to use a real API, and only the API.

**Comment by timotheecour** [comment]
> You simply cannot read from /dev/urandom and know what you're gonna get, it might just be a file of this name

meh, we can just add `enforce cfile.st_mode == S_ISCHR, "/dev/urandom is not of the expected kind (expecting character special), your system is broken"`

Which I've never heard of, seems like a hypothetical scenario. Again, this is a fallback, almost all systems (and all the OSes officially supported by nim) in use have APIs for that and don't need to /dev/urandom direcly (as can be seen in https://nim-lang.github.io/Nim/sysrand.html#urandom%2CNatural)

**Comment by Araq** [comment]
Just because it's "character special" that doesn't mean that it produces random numbers... ;-)

**Comment by c-blake** [comment]
If a system is so compromised then hard linking `/dev/zero` (in some sense the least random numbers) to `/dev/urandom` might not even be a surprising strategy and `/dev/zero` is a character device file.  Lol.  Yes, these poor folks are clearly screwed security-wise...It is true that this is just a fallback, but some perfect RN situation also seems off point.

One, probably slow source of true random numbers for seeding/crypto/other performance insensitive applications and one PRNG seem enough for the stdlib.  The application areas are distinct enough that I don't think a "true Gaussian" adds much value.  It is also unreasonable/out of scope for the stdlib to keep up with [all the probability distributions](https://en.wikipedia.org/wiki/List_of_probability_distributions) that might get requested.  So, I doubt we need layering/multisource RNs in the in-stdlib.

There are already not one but ***three*** nimble packages with their own implementations of `mersenne`, actually - [nim-random](https://github.com/oprypin/nim-random) and [nimscripter](https://github.com/beef331/nimscripter) and [libtcod](https://github.com/Vladar4/libtcod_nim/).  `alea` which allows layering backend sources via concepts could be made to pick one of those to depend upon (or add its own 4th!).

So, this PR could simply deprecate `mersenne`.  If @pyautogui wants to lift it into a 4th package that only provides just that then this also seems reasonable.  Then after a deprecation cycle, it can become deletion.

To help understand how much disturbance this might cause, I searched a clone of the nimbleverse I have.  The only packages I could find that seem to depend upon `mersenne` being in the stdlib are two with *example code only*:
```
desim http://github.com/jayvanderwall/desim
plotly https://github.com/brentp/nim-plotly
```
and two that *actually use it*:
```
perlin  https://github.com/Nycto/PerlinNim
  (just 1 line of code; easily changed)
opengl-sandbox  https://github.com/krux02/opengl-sandbox
  (this uses it the most - like 4 or 5 places)
```
There is who knows how much non-nimble "dark code" out there, of course..but only one real usage in 1500 packages suggests extremely weak reliance/low perturbation to deprecating now.

**Comment by qaziquza** [comment]
Yes, that sounds good. I am added {. deprecated .} pragmas currently.

File: lib/pure/mersenne.nim [hunk 1]
```diff
@@ -35,14 +36,16 @@ proc newMersenneTwister*(seed: uint32): MersenneTwister =
...
   for i in 0..623:
     var y = (m.mt[i] and 0x80000000'u32) +
             (m.mt[(i+1) mod 624] and 0x7fffffff'u32)
     m.mt[i] = m.mt[(i+397) mod 624] xor uint32(y shr 1'u32)
     if (y mod 2'u32) != 0:
       m.mt[i] = m.mt[i] xor 0x9908b0df'u32
 
-proc getNum*(m: var MersenneTwister): uint32 =
+proc getNum*(m: var MersenneTwister): uint32 {.deprecated:
+"deprecated, use random.rand() instead"} =
```
Comment (konsumlamm) [inline lib/pure/mersenne.nim]:
```suggestion
"use random.rand() instead"} =
```
No need to repeat that it's deprecated, the warning will say that anyway (same for the other annotations).

File: lib/pure/mersenne.nim [hunk 2]
```diff
@@ -53,3 +56,29 @@ proc getNum*(m: var MersenneTwister): uint32 =
...
+  let randNum = uint32(getNum(mt))
+  var largerNumIndex: uint32 = 0
+  var smallerNumIndex: uint32 = 1
+  while largerNumIndex < uint32(len(correspondingValues))  and smallerNumIndex < uint32(len(correspondingValues)):
+   if randNum < correspondingValues[largerNumIndex] and randNum > correspondingValues[smallerNumIndex]:
+    return arr[largerNumIndex]
+   elif randNum < correspondingValues[correspondingValues.high]:
+     return arr[correspondingValues.high]
+   largerNumIndex = largerNumIndex + 1
+   smallerNumIndex = smallerNumIndex + 1
```
Comment (konsumlamm) [inline lib/pure/mersenne.nim]: If the module is going to be deprecated anyway, what's the point in adding these functions?

**Comment by konsumlamm** [comment]
I'd also add `{.deprecated: "use std/random instead".}` at the top level, to deprecate the module itself and not only everything inside it.

**Comment by timotheecour** [comment]
> I'd also add

I'd only add. No need to deprecate individual procs, just the module

File: lib/pure/mersenne.nim (hunk 2 above)
Comment (qaziquza) [reply lib/pure/mersenne.nim]: I will remove them, sorry.

File: lib/pure/mersenne.nim [hunk 3]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (konsumlamm) [inline lib/pure/mersenne.nim]:
```suggestion
```

File: lib/pure/mersenne.nim [hunk 4]
```diff
@@ -10,7 +11,8 @@
 ## The [Mersenne Twister](https://en.wikipedia.org/wiki/Mersenne_Twister)
 ## random number generator.
 ##
-## **Note:** The procs in this module work at compile-time.
+## **Note:** The procs in this module work at compile-time
```
Comment (konsumlamm) [inline lib/pure/mersenne.nim]:
```suggestion
## .. note:: The procs in this module work at compile-time.
```

File: lib/pure/mersenne.nim [hunk 5]
```diff
@@ -35,6 +37,7 @@ proc newMersenneTwister*(seed: uint32): MersenneTwister =
                                       (result.mt[i-1] shr 30'u32)) + i)
 
 proc generateNumbers(m: var MersenneTwister) =
+
```
Comment (konsumlamm) [inline lib/pure/mersenne.nim:34]:
```suggestion
```

File: lib/pure/mersenne.nim [hunk 6]
```diff
@@ -52,4 +55,4 @@ proc getNum*(m: var MersenneTwister): uint32 =
   result = result xor (result shr 11'u32)
   result = result xor ((result shl 7'u32) and 0x9d2c5680'u32)
   result = result xor ((result shl 15'u32) and 0xefc60000'u32)
-  result = result xor (result shr 18'u32)
+  result = result xor (result shr 18'u32)
```
Comment (konsumlamm) [inline lib/pure/mersenne.nim]: Why remove the trailing newline? :(

File: tests/stdlib/tmersenne.nim [hunk 7]
```diff
@@ -6,6 +6,13 @@ template main() =
   doAssert mt.getNum == 407788156'u32
   doAssert mt.getNum == 1071751096'u32
   doAssert mt.getNum == 3805347140'u32
+  doAssert sample(mt, @[30, 20, 10]) == 20
+  doAssert sample(mt, @[30.4, 20.54, 10.5]) == 10.5
+  doAssert sample(mt, @['a', 'b', 'c']) == 'b'
+  doAssert sample(mt, @["abc", "def", "ghi"]) == "def"
+  
+
+
```
Comment (konsumlamm) [inline tests/stdlib/tmersenne.nim]: These tests won't compile without the proc.

Comment (qaziquza) [reply tests/stdlib/tmersenne.nim]: Sorry, stupid of me. Fixing.

**Comment by c-blake** [comment]
You should maybe change the title of this PR to "deprecate mersenne" or similar and downvoters should perhaps reconsider their votes. :-)  The whitespace changes seem a bit superfluous, but :shrug: also harmless.  Other than that, it's looking fine.

File: lib/pure/mersenne.nim [hunk 8]
```diff
@@ -2,24 +2,19 @@
...
 #    See the file "copying.txt", included in this
 #    distribution, for details about the copyright.
-#
-
 ## The [Mersenne Twister](https://en.wikipedia.org/wiki/Mersenne_Twister)
 ## random number generator.
-##
-## **Note:** The procs in this module work at compile-time.
+## .. note:: The procs in this module work at compile-time.
+{.deprecated:"import std/random instead.".}
```
Comment (ringabout) [inline lib/pure/mersenne.nim]: May need a changelog

Comment (timotheecour) [reply lib/pure/mersenne.nim]:
```suggestion

{.deprecated: "use `std/random` instead.".}
```

**Comment by Araq** [comment]
Needs a changelog entry.

File: changelog.md [hunk 9]
```diff
@@ -95,6 +94,7 @@
 
 
 ## Standard library additions and changes
+- Deprecated std/mersenne
```
Comment (timotheecour) [inline changelog.md]: this goes under: `## Changes affecting backward compatibility`

Comment (qaziquza) [reply changelog.md]: Just fixed it. Sorry.

File: changelog.md [hunk 10]
```diff
@@ -3,6 +3,7 @@
 
 
 ## Changes affecting backward compatibility
+- Deprecate std/mersenne
```
Comment (timotheecour) [inline changelog.md]:
```suggestion

- Deprecated `std/mersenne`

```

**Review by timotheecour — APPROVED** [review]

**Pull request merged**